test/validation/api/pktio/pktio.c

```diff
@@ -2578,6 +2578,202 @@ static void pktio_test_extra_stats(void)
 	CU_ASSERT(odp_pktio_close(pktio) == 0);
 }
 
+/* Per packet protocol statistics are platform specific,
+ * skip pktio tests when underline platform not supports
+ */
+static int pktio_check_proto_statistics_counters(void)
+{
+	odp_proto_stats_capability_t capa;
+	odp_pktio_param_t pktio_param;
+	odp_pktio_t pktio;
+	int ret;
+
+	odp_pktio_param_init(&pktio_param);
+	pktio_param.in_mode = ODP_PKTIN_MODE_SCHED;
+
+	pktio = odp_pktio_open(iface_name[0], pool[0], &pktio_param);
+	if (pktio == ODP_PKTIO_INVALID)
+		return ODP_TEST_INACTIVE;
+
+	ret = odp_proto_stats_capability(pktio, &capa);
+	(void)odp_pktio_close(pktio);
+
+	if (ret < 0 || capa.tx.counters.all_bits == 0)
+		return ODP_TEST_INACTIVE;
+
+	return ODP_TEST_ACTIVE;
+}
+
+static void pktio_test_proto_statistics_counters(void)
+{
+	odp_pktio_t pktio_rx, pktio_tx;
+	odp_pktio_t pktio[MAX_NUM_IFACES] = {
+		ODP_PKTIO_INVALID, ODP_PKTIO_INVALID
+	};
+	odp_packet_t pkt;
+	odp_packet_t tx_pkt[1000];
+	uint32_t pkt_seq[1000];
+	odp_event_t ev;
+	int i, pkts, tx_pkts, ret, alloc = 0;
+	odp_pktout_queue_t pktout;
+	uint64_t wait = odp_schedule_wait_time(ODP_TIME_MSEC_IN_NS);
+	int64_t flow0_pkts = 0, flow1_pkts = 0;
+	odp_proto_stats_capability_t capa;
+	odp_packet_proto_stats_opt_t opt0;
+	odp_packet_proto_stats_opt_t opt1;
+	odp_proto_stats_param_t param;
+	odp_proto_stats_data_t data0;
+	odp_proto_stats_data_t data1;
+	odp_pktio_config_t config;
+	odp_proto_stats_t stat0;
+	odp_proto_stats_t stat1;
+
+	memset(&pktout, 0, sizeof(pktout));
+	odp_pktio_config_init(&config);
+	config.pktout.bit.proto_stats_ena = 1;
+
+	for (i = 0; i < num_ifaces; i++) {
+		pktio[i] = create_pktio(i, ODP_PKTIN_MODE_SCHED,
+					ODP_PKTOUT_MODE_DIRECT);
+
+		CU_ASSERT_FATAL(pktio[i] != ODP_PKTIO_INVALID);
+		/* Enable protocol stats on interface 0 (Tx interface) */
+		if (!i)
+			odp_pktio_config(pktio[i], &config);
```

Comment:
There is no point in putting this inside the loop when it is done only for the first item. And the return value should be checked. So code like this after the loop and after assigning pktio_tx and pktio_rx would be better:

```
                /* Enable protocol stats Tx interface */
                odp_pktio_config_init(&config);
                config.pktout.bit.proto_stats_ena = 1;
                ret = odp_pktio_config(pktio_tx, &config);
                CU_ASSERT(ret == 0);
```

Reply:
Thanks for suggestion, will change in next version